lambdas/pkgpush/src/t4_lambda_pkgpush/__init__.py

```diff
@@ -228,138 +92,131 @@ def from_quilt_exception(cls, qe: quilt3.util.QuiltException):
             for entry in entries
         ]
         for f in concurrent.futures.as_completed(fs):
             f.result()
 
-
-def calculate_pkg_entry_hash(get_client_for_bucket, pkg_entry):
-    pk = pkg_entry.physical_key
-    params = {
-        'Bucket': pk.bucket,
-        'Key': pk.path,
-    }
-    if pk.version_id is not None:
-        params['VersionId'] = pk.version_id
-    url = get_client_for_bucket(pk.bucket).generate_presigned_url(
-        ClientMethod='get_object',
-        ExpiresIn=S3_HASH_LAMBDA_SIGNED_URL_EXPIRES_IN_SECONDS,
-        Params=params,
-    )
-    pkg_entry.hash = {
-        'type': 'SHA256',
-        'value': invoke_hash_lambda(url),
-    }
-
-
-def invoke_hash_lambda(url):
-    resp = lambda_.invoke(FunctionName=S3_HASH_LAMBDA, Payload=json.dumps(url))
-    if 'FunctionError' in resp:
-        raise PkgpushException('S3HashLambdaUnhandledError')
-    return json.load(resp['Payload'])
+    logger.debug("calculate_pkg_hashes: done")
 
 
 # Isolated for test-ability.
-get_user_boto_session = boto3.session.Session
+get_user_boto_session = boto3.Session
 
 
-@contextlib.contextmanager
-def setup_user_boto_session(session):
-    global user_boto_session
-    user_boto_session = session
-    try:
-        yield user_boto_session
-    finally:
-        user_boto_session = None
+class Event(pydantic.BaseModel):
+    credentials: AWSCredentials
+    params: T.Any
 
 
 def auth(f):
-    validator = Draft7Validator(CREDENTIALS_SCHEMA)
-
     @functools.wraps(f)
-    def wrapper(event):
-        credentials = event.get("credentials")
-        # TODO: collect all errors
-        ex = next(validator.iter_errors(credentials), None)
-        if ex is not None:
-            raise PkgpushException("InvalidCredentials", {"details": ex.message})
-
-        with setup_user_boto_session(get_user_boto_session(**credentials)):
-            return f(event.get("params"))
+    @pydantic.validate_arguments
+    def wrapper(event: Event):
+        session = get_user_boto_session(**event.credentials.boto_args)
+        token = user_boto_session.set(session)
+        try:
+            return f(event.params)
+        finally:
+            user_boto_session.reset(token)
+
     return wrapper
 
 
 def exception_handler(f):
     @functools.wraps(f)
-    def wrapper(event, _context):
+    def wrapper(event, context):
+        logger.debug("event: %s", event)
+        logger.debug("context: %s", context)
         try:
-            return {"result": f(event)}
+            result = f(event)
+            logger.debug("result: %s", result)
+            return {"result": result.dict()}
+        except RequestTooLarge as e:
+            logger.exception("RequestTooLarge")
+            return {
+                "error": {
+                    "name": "RequestTooLarge",
+                    "context": {
+                        "size": e.size,
+                        "max_size": e.max_size,
+                    },
+                }
+            }
```

Comment:
Might make sense to move that inside nested try and re-raise as PkgpushException.
Also might make sense to add types for result/error dicts especially if we're going to instantiate error dict explicitly.